### PR TITLE
InputControl: Ignore IME events when `isPressEnterToChange`

### PR DESCRIPTION
**Pull request by mirka**
Follow-up to #59081
Part of #45605

## What?

Ignore <kbd>Enter</kbd> and <kbd>Esc</kbd> keydowns emitted during IME composition when the `isPressEnterToChange` prop is enabled.

When this prop is enabled:

- <kbd>Enter</kbd> will need to be hit before an `onChange` event is fired. <kbd>Enter</kbd> keys pressed during IME composition are not intended to commit the change in the entire input field, so they should be ignored.
- <kbd>Esc</kbd> will clear the input field when there are "uncommitted" changes (i.e. changes that have not been committed with <kbd>Enter</kbd> yet). <kbd>Esc</kbd> keys pressed during IME composition are not intended to clear the entire input field, so they should be ignored.

## Why?

This was found during an audit of the components package to see if there is any keydown handling logic that doesn't ignore IME events properly. This was the only violation that I found. (There _is_ unfiltered keydown handling logic in other components too, but they are unlikely to involve IME events.)

## Testing Instructions

1. On the "Default" story for InputControl in the Storybook, enable the `isPressEnterToChange` prop and reload.
2. Type with your IME in the input field. <kbd>Enter</kbd> keypresses during composition should not log an `onChange` event in the Actions panel.
3. Reload. Add some text to the input field, but without hitting <kbd>Enter</kbd> to commit the changes to the InputControl. Type some more text using your IME and <kbd>Esc</kbd> before completing the composition. The existing text in the input field should remain, and only the uncompleted composition should be canceled.

### Before

https://github.com/WordPress/gutenberg/assets/555336/309f721b-4e12-4eb4-93a1-f7189d5e70c9

### After

https://github.com/WordPress/gutenberg/assets/555336/8d823015-2726-4da9-83a5-84abaddf8bfa

### CODE REVIEW

**Review by t-hamano — COMMENTED** [review]
I think it makes sense to add HOF, but strangely, I was unable to reproduce this problem in my environment (OS: Win11, browser: Chrome/Firefox).

In the video below, I add `console.log(key)` to observe the key down event of the `InputControl` component in trunk and start Storybook. During IME input, `event.key` is _always_ `"Process"`, including the `Enter` key and `Esc` key. In your environment, even during IME input, does `event.key` show the actual key?


https://github.com/WordPress/gutenberg/assets/54422211/63307fe9-277b-41db-a5f2-448c04646e70

**Review by tyxla — COMMENTED** [review]
Code-wise this is a straightforward change and looks good, but I'll defer to @t-hamano for actually verifying it works as expected.

**Review by t-hamano — APPROVED** [review]
Interestingly, `keyboardEvent.key` seems to be different depending on the OS. I would like to introduce two resources.

---

https://github.com/seawind543/keydown-key

> - With IME, the keyDown.key value of Chrome is different on Mac and Windows
>   - Mac: key === `Enter`
>   - Windows: key === `Process`
> - With IME, the keyDown.key value of Chrome and FireFox are different on Mac
>   - FireFox key === `Process` on both Mac and Windows

As stated above, the values seem to vary depending on the OS and browser.

---

https://youtrack.jetbrains.com/issue/PRJ-514/IME-isnt-supported-for-Mac

It is shown that the value of `event.key` differs depending on the OS.

---

If all of these resources are correct, this issue may only occur with the Chrome browser on Mac.

In any case, the changes in this PR have no effect on Windows, so LGTM 👍

**Review by torounit — APPROVED** [review]
I was able to reproduce this issue on macOS 14 / Chrome, and also confirmed that it does not occur in Firefox.

## bofore

https://github.com/WordPress/gutenberg/assets/1908815/6f7de97c-2492-4b3c-9797-3fbf61e64abd

## after

https://github.com/WordPress/gutenberg/assets/1908815/b0fbcf2e-be83-4038-ae73-39f8facbd9c7

**Comment by mirka** [comment]
> As stated above, the values seem to vary depending on the OS and browser.

😭

I'm glad we have both Win/Mac users for coverage though, thank you all for testing!

**Pull request merged**
